R/02-filter.R

```diff
@@ -61,30 +62,69 @@ qc_filter_zerocounts <- function(gimap_dataset){
 #'   qc_filter_plasmid(gimap_dataset, cutoff=2)
+#'   
+#'   #or to specify a different column (or set of columns to select)
+#'   qc_filter_plasmid(gimap_dataset, filter_plasmid_target_col = c(1,2))
+#'
+#'   # or to specify a cutoff value that will be used in the filter rather than the lower outlier default as well as to specify a different column (or set of columns) to select
+#'   qc_filter_plasmid(gimap_dataset, cutoff=1.75, filter_plasmid_target_col=c(1,2))
+#' 
 #' }
 #'
 
-qc_filter_plasmid <- function(gimap_dataset, cutoff = NULL){
-  plasmid_data <- data.frame(gimap_dataset$transformed_data$log2_cpm[, 1]) %>% `colnames<-`(c("log2_cpm"))
+qc_filter_plasmid <- function(gimap_dataset, cutoff = NULL, filter_plasmid_target_col = NULL){
+  
+  if (is.null(filter_plasmid_target_col)) {filter_plasmid_target_col <- c(1)}
+  
+  plasmid_data <- data.frame(gimap_dataset$transformed_data$log2_cpm[, filter_plasmid_target_col]) %>% `colnames<-`(rep(c("plasmid_log2_cpm"), length(filter_plasmid_target_col))) %>% clean_names()
+  
+  if (length(filter_plasmid_target_col >1)){ #if more than one column was selected, collapse all of the columns into the same vector using pivot_longer to store in a df with the name of the rep and number for row/construct
+    plasmid_data <- plasmid_data %>% mutate(construct = rownames(plasmid_data)) %>%
+      pivot_longer(starts_with("plasmid_log2_cpm"), 
+                   values_to = "plasmid_log2_cpm", 
+                   names_to = "rep")
+  }
   
   if (is.null(cutoff)) {
     # if cutoff is null, use lower outlier 
-    quantile_info <- quantile(plasmid_data$log2_cpm)
+    quantile_info <- quantile(plasmid_data$plasmid_log2_cpm)
     
     cutoff <- quantile_info["25%"] - (1.5 * (quantile_info["75%"] - quantile_info["25%"])) #later step make a function for this in utils since it's used more than once
   }
   
-  plasmid_cpm_filter <- unlist(lapply(1:nrow(plasmid_data), function(x) plasmid_data$log2_cpm[x] < cutoff))
+  if (length(filter_plasmid_target_col >1)){ #if more than one column was selected, take collapsed/pooled data and compare it to the cutoff
+                                            #then pivot_wider so that the constructs are in the same row and we can use if_any to report if any of the replicates were flagged by the cutoff
+                                            #return just that summary column (reporting if any are TRUE) as the filter
+    plasmid_data <- plasmid_data %>% 
+      mutate(filterFlag = plasmid_log2_cpm < cutoff) %>%
+      pivot_wider(id_cols = construct, names_from = rep, values_from = filterFlag)
+    plasmid_cpm_filter <- plasmid_data %>% 
+      mutate(plasmid_cpm_filter=  if_any(.cols = starts_with('plasmid_log2_cpm'))) %>%
+      select(plasmid_cpm_filter)
+    
+  } else {
+  
+    plasmid_cpm_filter <- as.data.frame(plasmid_data$plasmid_log2_cpm < cutoff) %>%`colnames<-`("plasmid_cpm_filter")
```

Comment:
```suggestion
    plasmid_cpm_filter <- as.data.frame(plasmid_data$plasmid_log2_cpm < cutoff)
    colnames<-("plasmid_cpm_filter")
```

Not sure much is added by doing the %>% with colnames. Can make these their own steps for readability. But I also don't feel super strongly about this if there's a reason you have it this way.